### ISSUE TITLE
Convert autoresearch-embed from PyTorch MPS to native MLX

### ISSUE THREAD

**Issue by pierretokns** [pierretokns/agenthub#13]
## Why We're Converting

The current training pipeline uses PyTorch with the MPS backend on an M2 Ultra (64GB). While functional, MPS introduces fundamental performance bottlenecks that MLX eliminates by design:

### PyTorch MPS Bottlenecks (measured)
- **CPU↔GPU synchronization**: Every `.to(device)`, `.cpu()`, and `.numpy()` call triggers a Metal command buffer flush + sync. Our encode() loop does this per batch (~hundreds of times per eval).
- **No tensor cores / no Flash Attention**: MPS translates CUDA-first kernels to Metal. No fused attention, no hardware matmul acceleration beyond basic Metal compute.
- **Eager execution overhead**: Each PyTorch op launches a separate Metal kernel with dispatch overhead. No graph fusion.
- **Padding waste**: Variable-length sequences padded to max batch length. At batch_size=256, the contrastive stage runs 31s/step vs 1.7s/step for fine-tuning (smaller batches).
- **Memory overhead**: Peak 79.5GB (over our 55GB target), requiring `torch.mps.empty_cache()` + `gc.collect()` between stages.

### What MLX Gives Us
- **Zero-copy unified memory**: MLX arrays live directly in unified memory. No CPU↔GPU transfer — `np.array(mx_array)` is a pointer cast, not a copy.
- **Native Metal kernels**: Purpose-built for Apple Silicon GPU, not a CUDA translation layer.
- **Lazy evaluation + graph fusion**: MLX fuses operations (matmul → add → gelu → normalize) into single Metal kernels. Dramatically reduces kernel launch overhead.
- **`mx.fast.scaled_dot_product_attention()`**: Native Metal attention — closest thing to Flash Attention on Apple Silicon.
- **Simpler code**: No `.to(device)`, no `autocast`, no `empty_cache()`, no `GradScaler`. Just arrays.

### Existing MLX Assets
Most of the conversion is already solved by community work:

| Component | Status | Source |
|-----------|--------|--------|
| ModernBERT weights in MLX | **DONE** | [mlx-community/tasksource-ModernBERT-base-embed-bf16](https://huggingface.co/mlx-community/tasksource-ModernBERT-base-embed-bf16) |
| ModernBERT encoder in MLX | **DONE** | [pappitti/modernbert-mlx](https://github.com/pappitti/modernbert-mlx) + [Blaizzy/mlx-embeddings](https://github.com/Blaizzy/mlx-embeddings) |
| Loss functions (InfoNCE etc.) | **DONE** | `src/losses.py` — already written in MLX (`import mlx.core as mx`) |
| MLX training loop patterns | **DONE** | `autoresearch-mlx/train.py` — our own repo has `nn.value_and_grad`, custom AdamW, grad accumulation |
| MLX contrastive retrieval training | **Reference** | [jina-ai/mlx-retrieval](https://github.com/jina-ai/mlx-retrieval) |
| MTEB eval interface | **Ready** | `src/eval/mteb_runner.py` already expects `encode_sentences() → numpy` |
| Data pipeline | **No change** | Framework-agnostic (returns list of dicts) |
| Decontamination | **No change** | Pure datasketch/numpy |

### What Needs to Be Written

1. **`EmbeddingModel` as `mlx.nn.Module`**: Load ModernBERT encoder from mlx-community weights, add mean pooling + projection + L2 normalize. Implement `encode_sentences()` returning numpy.
2. **Training loop rewrite**: Replace PyTorch backward/step with `nn.value_and_grad()` pattern (same as autoresearch-mlx). Remove all `.to(device)`, autocast, empty_cache.
3. **Optimizer**: Use `mlx.optimizers.AdamW` or port per-parameter AdamW from autoresearch-mlx for differential learning rates.
4. **Hard negative mining**: Replace `torch.matmul` + `torch.topk` with `mx.matmul` + numpy sorting.
5. **Checkpoint save/load**: Switch from `torch.save` to safetensors.

### Expected Performance Impact
- **Memory**: From 79GB peak → ~35-40GB (no duplication between CPU/GPU)
- **Training throughput**: 2-4x improvement expected (native Metal, no sync overhead, fused ops)
- **Eval speed**: Significant improvement in encode() — no per-batch CPU↔GPU copies
- **Code simplicity**: ~200 fewer lines (remove all PyTorch device management boilerplate)

### Current Best Result (PyTorch MPS)
```
primary_score: 27.20  (only clean/kept experiment)
STSBenchmark: 45.71, SICK-R: 49.23, TwitterURLCorpus: 68.08
SprintDuplicateQuestions: 34.12, TwentyNewsgroupsClustering: 9.11
RedditClustering: 8.82, SciFact: 4.06, NFCorpus: 2.25
```

### Experiments run: 5 (2 contaminated, 1 kept, 1 discard, 1 in-progress killed)

The MLX conversion should let us run significantly more experiments per hour, accelerating the research loop toward our MTEB targets.

### References
- [MLX Framework](https://github.com/ml-explore/mlx)
- [MLX Community ModernBERT Collection](https://huggingface.co/collections/mlx-community/modernbert-67eea1cbffcd7f6fd1c09217)
- [pappitti/modernbert-mlx](https://github.com/pappitti/modernbert-mlx)
- [Blaizzy/mlx-embeddings](https://github.com/Blaizzy/mlx-embeddings)
- [jina-ai/mlx-retrieval](https://github.com/jina-ai/mlx-retrieval)
- [WWDC25: Get started with MLX](https://developer.apple.com/videos/play/wwdc2025/315/)
